File: backend/DataFileUnification.py
import pandas as pd
import numpy as np
from pathlib import Path
import sys
import os

def find_required_files(folder_path):
    """Find the 3 required CSV files in the folder"""
    folder_path = Path(folder_path)
    
    if not folder_path.exists():
        raise FileNotFoundError(f"Folder does not exist: {folder_path}")
    
    required_files = {
        "1HZLOG.CSV": None,
        "10HZLOG.CSV": None,
        "100HZLOG.CSV": None
    }
    
    # Search for files (case-insensitive)
    for file_path in folder_path.glob("*.csv"):
        file_name = file_path.name.upper()
        for required_file in required_files.keys():
            if file_name == required_file:
                required_files[required_file] = file_path
                break
    
    # Also try .CSV extension
    for file_path in folder_path.glob("*.CSV"):
        file_name = file_path.name.upper()
        for required_file in required_files.keys():
            if file_name == required_file:
                required_files[required_file] = file_path
                break
    
    # Check if all files were found
    missing_files = [name for name, path in required_files.items() if path is None]
    if missing_files:
        raise FileNotFoundError(f"Missing required files: {', '.join(missing_files)}")
    
    return (
        required_files["100HZLOG.CSV"],
        required_files["10HZLOG.CSV"], 
        required_files["1HZLOG.CSV"]
    )

def detect_header_format(file_path):
    """Detect if the file has extended headers (4 lines) or simple headers (1 line)"""
    with open(file_path, 'r') as file:
        lines = file.readlines()
    
    if len(lines) < 4:
        return False, lines[0].strip()
    
    header_line = lines[0].strip()
    
    # Simple heuristic: try to parse second line as data
    try:
        parts = lines[1].split(',')
        int(parts[0].strip())  # Should be a timestamp
        return False, header_line
    except (ValueError, IndexError):
        return True, header_line

def clean_csv_data(file_path):
    """Clean CSV data by removing header repetitions and handling time restarts"""
    has_extended_headers, header_line = detect_header_format(file_path)
    header_lines_count = 4 if has_extended_headers else 1
    
    with open(file_path, 'r') as file:
        lines = file.readlines()
    
    # Get expected column count from header
    expected_columns = len(header_line.split(','))
    
    # Create a standardized 4-line header
    if has_extended_headers:
        # Use existing extended header
        complete_header = ''.join(lines[:4])
    else:
        # Create 4-line header from single line
        column_names = header_line.split(',')
        units_line = ','.join(['unknown'] * len(column_names))
        conversion_line = ','.join(['-7'] * len(column_names))
        precision_line = ','.join(['32'] * len(column_names))
        
        complete_header = f"{header_line}\n{units_line}\n{conversion_line}\n{precision_line}\n"
    
    # Filter out header repetitions and collect data
    data_lines = []
    last_time = 0
    time_offset = 0
    
    i = header_lines_count  # Start after the initial header
    while i < len(lines):
        line = lines[i].strip()
        
        # Check if this is a header repetition
        if line == header_line:
            i += header_lines_count
            continue
        
        # Process data line
        if line:
            parts = line.split(',')
            
            # Fix malformed lines with extra data points
            if len(parts) > expected_columns:
                parts = parts[:expected_columns]  # Truncate to expected length
            
            if len(parts) >= 2:
                try:
                    current_time = int(parts[0])
                    
                    # Apply current time offset to get the adjusted time
                    adjusted_time = current_time + time_offset
                    
                    # Check if time has restarted
                    if adjusted_time < last_time:
                        time_offset = last_time
                        adjusted_time = current_time + time_offset
                    
                    # Update the parts with the adjusted time
                    parts[0] = str(adjusted_time)
                    
                    # Ensure we have the right number of columns
                    if len(parts) == expected_columns:
                        data_lines.append(','.join(parts))
                        last_time = adjusted_time
                except ValueError:
                    pass  # Skip malformed lines
        
        i += 1
    
    # Write cleaned data to temporary file
    cleaned_file = file_path.with_suffix('.cleaned.csv')
    with open(cleaned_file, 'w') as file:
        # Always write the complete 4-line header
        file.write(complete_header)
        file.write('\n'.join(data_lines))
    
    return cleaned_file

def interpolate_data(source_df, target_times, time_col='Time'):
    """Interpolate data from source DataFrame to target times"""
    # Find the actual time column name
    actual_time_col = None
    for col in source_df.columns:
        if col.strip().lower() == time_col.lower():
            actual_time_col = col
            break
    
    if actual_time_col is None:
        raise ValueError(f"Time column '{time_col}' not found in source data")
    
    # Create result DataFrame with target times
    result_df = pd.DataFrame({time_col: target_times})
    
    # Interpolate each column except Time and GLOBAL time
    for col in source_df.columns:
        if col != actual_time_col and 'global' not in col.lower():
            result_df[col] = np.interp(target_times, source_df[actual_time_col], source_df[col])
    
    return result_df

def combine_data_files(hz_100_file, hz_10_file, hz_1_file, output_file):
    """Combine three CSV files with different sampling rates into one file"""
    # Convert to Path objects
    hz_100_file = Path(hz_100_file)
    hz_10_file = Path(hz_10_file)
    hz_1_file = Path(hz_1_file)
    
    # Clean the data files
    cleaned_100 = clean_csv_data(hz_100_file)
    cleaned_10 = clean_csv_data(hz_10_file)
    cleaned_1 = clean_csv_data(hz_1_file)
    
    try:
        # Read the header information from the cleaned files
        def read_header_info(file_path):
            with open(file_path, 'r') as f:
                lines = f.readlines()
                return {
                    'columns': lines[0].strip().split(','),
                    'units': lines[1].strip().split(','),
                    'conversion': lines[2].strip().split(','),
                    'precision': lines[3].strip().split(',')
                }
        
        header_100 = read_header_info(cleaned_100)
        header_10 = read_header_info(cleaned_10)
        header_1 = read_header_info(cleaned_1)
        
        # Read cleaned data - always skip the first 3 lines (units, conversion, precision)
        # since we now always generate 4-line headers
        df_100 = pd.read_csv(cleaned_100, skiprows=[1, 2, 3])
        df_10 = pd.read_csv(cleaned_10, skiprows=[1, 2, 3])
        df_1 = pd.read_csv(cleaned_1, skiprows=[1, 2, 3])
        
        # Remove GLOBAL time columns
        for df in [df_100, df_10, df_1]:
            global_time_cols = [col for col in df.columns if 'global' in col.lower()]
            for col in global_time_cols:
                df.drop(col, axis=1, inplace=True)
        
        # Find time column
        time_col = None
        for col in df_100.columns:
            if col.strip().lower() == 'time':
                time_col = col
                break
        
        if time_col is None:
            raise ValueError("Time column not found in 100Hz data")
        
        target_times = df_100[time_col].values
        
        # Interpolate data
        df_10_interp = interpolate_data(df_10, target_times, time_col)
        df_1_interp = interpolate_data(df_1, target_times, time_col)
        
        # Combine all data
        combined_df = df_100.copy()
        
        # Add columns from other files
        for col in df_10_interp.columns:
            if col.strip().lower() != 'time':
                combined_df[col] = df_10_interp[col]
        
        for col in df_1_interp.columns:
            if col.strip().lower() != 'time':
                combined_df[col] = df_1_interp[col]
        
        # Create the complete 4-line header for the output file
        # Use existing header information where available
        final_columns = list(combined_df.columns)
        final_units = []
        final_conversion = []
        final_precision = []
        
        for col in final_columns:
            # Check if this column exists in any of the source headers
            col_found = False
            
            # Check 100Hz file first
            if col in header_100['columns']:
                idx = header_100['columns'].index(col)
                final_units.append(header_100['units'][idx])
                final_conversion.append(header_100['conversion'][idx])
                final_precision.append(header_100['precision'][idx])
                col_found = True
            # Check 10Hz file
            elif col in header_10['columns']:
                idx = header_10['columns'].index(col)
                final_units.append(header_10['units'][idx])
                final_conversion.append(header_10['conversion'][idx])
                final_precision.append(header_10['precision'][idx])
                col_found = True
            # Check 1Hz file
            elif col in header_1['columns']:
                idx = header_1['columns'].index(col)
                final_units.append(header_1['units'][idx])
                final_conversion.append(header_1['conversion'][idx])
                final_precision.append(header_1['precision'][idx])
                col_found = True
            
            # If not found in any source, use defaults
            if not col_found:
                final_units.append('unknown')
                final_conversion.append('-7')
                final_precision.append('32')
        
        # Write to output file with complete 4-line header
        with open(output_file, 'w', newline='') as f:
            f.write(','.join(final_columns) + '\n')
            f.write(','.join(final_units) + '\n')
            f.write(','.join(final_conversion) + '\n')
            f.write(','.join(final_precision) + '\n')
            
            # Write the data without extra newlines
            combined_df.to_csv(f, index=False, header=False, lineterminator='\n')
        
    finally:
        # Clean up temporary files
        for temp_file in [cleaned_100, cleaned_10, cleaned_1]:
            if temp_file.exists():
                temp_file.unlink()

def process_directory(directory_path):
    """Main function to process a directory and create fullData.csv"""
    try:
        # Find required files
        hz_100_file, hz_10_file, hz_1_file = find_required_files(directory_path)
        
        # Set output file path
        output_file = Path(directory_path) / "fullData.csv"
        
        # Combine the files
        combine_data_files(hz_100_file, hz_10_file, hz_1_file, output_file)
        
        return 0  # Success
        
    except Exception as e:
        print(f"Error: {str(e)}", file=sys.stderr)
        return 1  # Error

def main():
    """Command line interface"""
    if len(sys.argv) != 2:
        print("Usage: python DataFileUnification.py <directory_path>", file=sys.stderr)
        return 1
    
    directory_path = sys.argv[1]
    return process_directory(directory_path)

if __name__ == "__main__":
    sys.exit(main())

File: backend/DataFileUnification.go
package Backend

import (
	"encoding/csv"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
	"sync"
)

// findRequiredFiles searches for the 3 required CSV files (case-insensitive)
func findRequiredFiles(folderPath string) (hz100, hz10, hz1 string, err error) {
	requiredFiles := map[string]*string{
		"1HZLOG.CSV":   &hz1,
		"10HZLOG.CSV":  &hz10,
		"100HZLOG.CSV": &hz100,
	}

	entries, err := os.ReadDir(folderPath)
	if err != nil {
		return "", "", "", fmt.Errorf("failed to read directory: %w", err)
	}

	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		upperName := strings.ToUpper(entry.Name())
		if ptr, exists := requiredFiles[upperName]; exists {
			*ptr = filepath.Join(folderPath, entry.Name())
		}
	}

	missing := []string{}
	for name, ptr := range requiredFiles {
		if *ptr == "" {
			missing = append(missing, name)
		}
	}
	if len(missing) > 0 {
		return "", "", "", fmt.Errorf("missing required files: %s", strings.Join(missing, ", "))
	}

	return hz100, hz10, hz1, nil
}

type headerInfo struct {
	columns    []string
	units      []string
	conversion []string
	precision  []string
}

// parsedFile is the column-major in-memory representation of a cleaned CSV.
// values[col] is a []float64 of length len(times) with that column's parsed data.
// times is the parsed time column (same length as each values[col]).
type parsedFile struct {
	header     headerInfo
	times      []int64
	values     [][]float64 // values[colIdx][rowIdx]
	timeColIdx int
}

// detectHeaderFormat checks if file has 4-line extended headers or 1-line simple headers.
// Returns hasExtended, headerLine (joined), and an open csv.Reader positioned just past
// the headers — caller can stream the rest.
func detectHeaderFormat(filePath string) (bool, []string, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return false, nil, err
	}
	defer file.Close()

	reader := csv.NewReader(file)
	lines := [][]string{}
	for i := 0; i < 4; i++ {
		line, err := reader.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return false, nil, err
		}
		lines = append(lines, line)
	}

	if len(lines) < 2 {
		return false, nil, fmt.Errorf("file too short")
	}

	headerLine := lines[0]

	// Check if line 2 is a duplicate of line 1
	if len(lines[1]) == len(headerLine) {
		isDuplicate := true
		for i := range headerLine {
			if lines[1][i] != headerLine[i] {
				isDuplicate = false
				break
			}
		}
		if isDuplicate {
			return false, headerLine, nil
		}
	}

	// Try to parse second line as numeric data — simple header
	if _, err := strconv.ParseInt(lines[1][0], 10, 64); err == nil {
		return false, headerLine, nil
	}

	// Second line is not numeric, assume extended header
	return true, headerLine, nil
}

// parseFileColumnMajor reads a CSV, cleans header repetitions and time restarts,
// and returns a column-major float64 representation. This replaces the old
// cleanCSVData → interpolateToTimeline pipeline that parsed every value twice.
func parseFileColumnMajor(filePath string) (*parsedFile, error) {
	hasExtended, headerCols, err := detectHeaderFormat(filePath)
	if err != nil {
		return nil, err
	}
	expectedColumns := len(headerCols)

	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	reader := csv.NewReader(file)
	reader.FieldsPerRecord = -1 // tolerate variable-width rows
	reader.ReuseRecord = true   // avoids per-row allocation; we copy what we keep

	// Skip the header lines we already read in detectHeaderFormat.
	headerLinesCount := 1
	if hasExtended {
		headerLinesCount = 4
	}

	// Read header lines from this fresh reader.
	header := headerInfo{
		columns:    headerCols,
		units:      make([]string, expectedColumns),
		conversion: make([]string, expectedColumns),
		precision:  make([]string, expectedColumns),
	}

	if hasExtended {
		for i := 0; i < headerLinesCount; i++ {
			line, err := reader.Read()
			if err != nil {
				return nil, fmt.Errorf("failed to read header line %d: %w", i, err)
			}
			switch i {
			case 1:
				copy(header.units, line)
			case 2:
				copy(header.conversion, line)
			case 3:
				copy(header.precision, line)
			}
		}
	} else {
		// Consume first line; defaults for the missing rows.
		if _, err := reader.Read(); err != nil {
			return nil, err
		}
		for i := 0; i < expectedColumns; i++ {
			header.units[i] = "unknown"
			header.conversion[i] = "-7"
			header.precision[i] = "32"
		}
	}

	timeColIdx := findTimeColumn(headerCols)
	if timeColIdx == -1 {
		return nil, fmt.Errorf("time column not found in %s", filepath.Base(filePath))
	}

	// Pre-grow column slices; we'll trim later.
	values := make([][]float64, expectedColumns)
	for j := range values {
		values[j] = make([]float64, 0, 1024)
	}
	times := make([]int64, 0, 1024)

	var (
		lastTime   int64 = 0
		timeOffset int64 = 0
		duplicateHeaderChecked = !hasExtended // simple-header files may have a duplicate row
	)

	for {
		record, err := reader.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, err
		}
		if len(record) == 0 {
			continue
		}

		// Skip a duplicate of the header on the first iteration after a simple header.
		if !duplicateHeaderChecked {
			duplicateHeaderChecked = true
			if len(record) == expectedColumns {
				dup := true
				for i := range record {
					if record[i] != headerCols[i] {
						dup = false
						break
					}
				}
				if dup {
					continue
				}
			}
		}

		// Skip header repetitions sprinkled mid-file.
		if len(record) == expectedColumns && record[timeColIdx] == headerCols[timeColIdx] {
			equalsHeader := true
			for i := range record {
				if record[i] != headerCols[i] {
					equalsHeader = false
					break
				}
			}
			if equalsHeader {
				if hasExtended {
					for k := 1; k < headerLinesCount; k++ {
						if _, err := reader.Read(); err != nil {
							goto finishedScanning
						}
					}
				}
				continue
			}
		}

		// Truncate over-long rows.
		if len(record) > expectedColumns {
			record = record[:expectedColumns]
		}
		if len(record) != expectedColumns {
			continue
		}

		currentTime, err := strconv.ParseInt(record[timeColIdx], 10, 64)
		if err != nil {
			continue
		}
		adjustedTime := currentTime + timeOffset
		if adjustedTime < lastTime {
			timeOffset = lastTime
			adjustedTime = currentTime + timeOffset
		}

		times = append(times, adjustedTime)
		lastTime = adjustedTime

		for j := 0; j < expectedColumns; j++ {
			if j == timeColIdx {
				values[j] = append(values[j], float64(adjustedTime))
				continue
			}
			s := record[j]
			if s == "" {
				values[j] = append(values[j], 0)
				continue
			}
			v, err := strconv.ParseFloat(s, 64)
			if err != nil {
				values[j] = append(values[j], 0)
				continue
			}
			values[j] = append(values[j], v)
		}
	}
finishedScanning:

	return &parsedFile{
		header:     header,
		times:      times,
		values:     values,
		timeColIdx: timeColIdx,
	}, nil
}

// dropGlobalColumns filters out columns whose names contain "global".
// Operates on the column-major representation (no row copies).
func dropGlobalColumns(p *parsedFile) {
	keep := make([]int, 0, len(p.header.columns))
	for i, col := range p.header.columns {
		if !strings.Contains(strings.ToLower(col), "global") {
			keep = append(keep, i)
		}
	}
	if len(keep) == len(p.header.columns) {
		return
	}

	newCols := make([]string, len(keep))
	newUnits := make([]string, len(keep))
	newConv := make([]string, len(keep))
	newPrec := make([]string, len(keep))
	newValues := make([][]float64, len(keep))

	for newIdx, oldIdx := range keep {
		newCols[newIdx] = p.header.columns[oldIdx]
		newUnits[newIdx] = p.header.units[oldIdx]
		newConv[newIdx] = p.header.conversion[oldIdx]
		newPrec[newIdx] = p.header.precision[oldIdx]
		newValues[newIdx] = p.values[oldIdx]
	}
	p.header.columns = newCols
	p.header.units = newUnits
	p.header.conversion = newConv
	p.header.precision = newPrec
	p.values = newValues
	p.timeColIdx = findTimeColumn(newCols)
}

// interpolateColumnMajor maps every non-time column of src onto targetTimes via
// linear interpolation. Columns are independent, so we parallelize across them.
func interpolateColumnMajor(src *parsedFile, targetTimes []int64) [][]float64 {
	numCols := len(src.values)
	out := make([][]float64, numCols)
	if len(src.times) == 0 {
		for j := range out {
			out[j] = make([]float64, len(targetTimes))
		}
		return out
	}

	// Pre-compute, for each target time, the source index "left of" it.
	// Single linear pass since both arrays are sorted.
	leftIdx := make([]int, len(targetTimes))
	{
		i := 0
		for t := 0; t < len(targetTimes); t++ {
			for i < len(src.times)-1 && src.times[i+1] < targetTimes[t] {
				i++
			}
			leftIdx[t] = i
		}
	}

	var wg sync.WaitGroup
	workers := runtime.NumCPU()
	if workers > numCols {
		workers = numCols
	}
	if workers < 1 {
		workers = 1
	}
	jobs := make(chan int, numCols)
	for j := 0; j < numCols; j++ {
		jobs <- j
	}
	close(jobs)

	for w := 0; w < workers; w++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for j := range jobs {
				col := make([]float64, len(targetTimes))
				if j == src.timeColIdx {
					for t := range targetTimes {
						col[t] = float64(targetTimes[t])
					}
					out[j] = col
					continue
				}
				srcVals := src.values[j]
				srcTimes := src.times
				lastSrc := len(srcTimes) - 1
				for t, target := range targetTimes {
					if target <= srcTimes[0] {
						col[t] = srcVals[0]
						continue
					}
					if target >= srcTimes[lastSrc] {
						col[t] = srcVals[lastSrc]
						continue
					}
					i := leftIdx[t]
					t1 := srcTimes[i]
					t2 := srcTimes[i+1]
					if t2 == t1 {
						col[t] = srcVals[i]
						continue
					}
					v1 := srcVals[i]
					v2 := srcVals[i+1]
					col[t] = v1 + (v2-v1)*float64(target-t1)/float64(t2-t1)
				}
				out[j] = col
			}
		}()
	}
	wg.Wait()
	return out
}

// findTimeColumn finds the "Time" column (case-insensitive)
func findTimeColumn(columns []string) int {
	for i, col := range columns {
		if strings.ToLower(strings.TrimSpace(col)) == "time" {
			return i
		}
	}
	return -1
}

// unifiedResult is the in-memory result of parsing+merging the 3 CSVs.
// Columns are concatenated 100Hz + 10Hz(no time) + 1Hz(no time).
type unifiedResult struct {
	header headerInfo
	values [][]float64 // column-major; values[col][row]
	rows   int
}

// processDirectoryInMemory does the complete CSV unification pipeline in memory,
// parallelizing the three file reads and the per-column interpolation.
// No intermediate CSV is written.
func processDirectoryInMemory(directoryPath string) (*unifiedResult, error) {
	hz100, hz10, hz1, err := findRequiredFiles(directoryPath)
	if err != nil {
		return nil, err
	}

	// Stage 1: parse all three files in parallel.
	type parseResult struct {
		idx  int
		data *parsedFile
		err  error
	}
	results := make(chan parseResult, 3)
	paths := [3]string{hz100, hz10, hz1}
	var wg sync.WaitGroup
	for i, p := range paths {
		wg.Add(1)
		go func(i int, p string) {
			defer wg.Done()
			data, err := parseFileColumnMajor(p)
			results <- parseResult{i, data, err}
		}(i, p)
	}
	wg.Wait()
	close(results)

	parsed := [3]*parsedFile{}
	for r := range results {
		if r.err != nil {
			return nil, fmt.Errorf("failed to parse %s: %w", filepath.Base(paths[r.idx]), r.err)
		}
		parsed[r.idx] = r.data
	}

	// Stage 2: drop global-time columns from each (cheap, in-place).
	dropGlobalColumns(parsed[0])
	dropGlobalColumns(parsed[1])
	dropGlobalColumns(parsed[2])

	data100, data10, data1 := parsed[0], parsed[1], parsed[2]

	if data100.timeColIdx == -1 {
		return nil, fmt.Errorf("time column missing in 100Hz data after cleanup")
	}
	if data10.timeColIdx == -1 {
		return nil, fmt.Errorf("time column missing in 10Hz data after cleanup")
	}
	if data1.timeColIdx == -1 {
		return nil, fmt.Errorf("time column missing in 1Hz data after cleanup")
	}

	targetTimes := data100.times

	// Stage 3: interpolate 10Hz and 1Hz against the 100Hz timeline in parallel.
	var interp10, interp1 [][]float64
	var stage3wg sync.WaitGroup
	stage3wg.Add(2)
	go func() {
		defer stage3wg.Done()
		interp10 = interpolateColumnMajor(data10, targetTimes)
	}()
	go func() {
		defer stage3wg.Done()
		interp1 = interpolateColumnMajor(data1, targetTimes)
	}()
	stage3wg.Wait()

	// Stage 4: assemble unified column-major result.
	combined := headerInfo{}
	combinedVals := make([][]float64, 0, len(data100.values)+len(data10.values)+len(data1.values))

	combined.columns = append(combined.columns, data100.header.columns...)
	combined.units = append(combined.units, data100.header.units...)
	combined.conversion = append(combined.conversion, data100.header.conversion...)
	combined.precision = append(combined.precision, data100.header.precision...)
	combinedVals = append(combinedVals, data100.values...)

	for j, col := range data10.header.columns {
		if strings.ToLower(strings.TrimSpace(col)) == "time" {
			continue
		}
		combined.columns = append(combined.columns, col)
		combined.units = append(combined.units, data10.header.units[j])
		combined.conversion = append(combined.conversion, data10.header.conversion[j])
		combined.precision = append(combined.precision, data10.header.precision[j])
		combinedVals = append(combinedVals, interp10[j])
	}
	for j, col := range data1.header.columns {
		if strings.ToLower(strings.TrimSpace(col)) == "time" {
			continue
		}
		combined.columns = append(combined.columns, col)
		combined.units = append(combined.units, data1.header.units[j])
		combined.conversion = append(combined.conversion, data1.header.conversion[j])
		combined.precision = append(combined.precision, data1.header.precision[j])
		combinedVals = append(combinedVals, interp1[j])
	}

	return &unifiedResult{
		header: combined,
		values: combinedVals,
		rows:   len(targetTimes),
	}, nil
}

// writeUnifiedCSV writes the unified result to fullData.csv as a cache for next load.
// Writes to a temp file then renames so concurrent readers never see a partial file.
func writeUnifiedCSV(outputPath string, u *unifiedResult) error {
	tmpPath := outputPath + ".part"
	outFile, err := os.Create(tmpPath)
	if err != nil {
		return fmt.Errorf("failed to create output file: %w", err)
	}

	writer := csv.NewWriter(outFile)

	if err := writer.Write(u.header.columns); err != nil {
		outFile.Close()
		os.Remove(tmpPath)
		return err
	}
	if err := writer.Write(u.header.units); err != nil {
		outFile.Close()
		os.Remove(tmpPath)
		return err
	}
	if err := writer.Write(u.header.conversion); err != nil {
		outFile.Close()
		os.Remove(tmpPath)
		return err
	}
	if err := writer.Write(u.header.precision); err != nil {
		outFile.Close()
		os.Remove(tmpPath)
		return err
	}

	row := make([]string, len(u.values))
	for r := 0; r < u.rows; r++ {
		for c := range u.values {
			row[c] = strconv.FormatFloat(u.values[c][r], 'f', -1, 64)
		}
		if err := writer.Write(row); err != nil {
			outFile.Close()
			os.Remove(tmpPath)
			return err
		}
	}
	writer.Flush()
	if err := writer.Error(); err != nil {
		outFile.Close()
		os.Remove(tmpPath)
		return err
	}
	if err := outFile.Close(); err != nil {
		os.Remove(tmpPath)
		return err
	}
	return os.Rename(tmpPath, outputPath)
}

// ProcessDirectory is the legacy entry point — finds files and writes fullData.csv.
// Retained for any external callers; uses the same in-memory pipeline.
func ProcessDirectory(directoryPath string) error {
	u, err := processDirectoryInMemory(directoryPath)
	if err != nil {
		return err
	}
	return writeUnifiedCSV(filepath.Join(directoryPath, "fullData.csv"), u)
}
